metadata-ingestion/src/datahub/ingestion/graph/client.py

```diff
@@ -1205,7 +1218,97 @@ def close(self) -> None:
+    config_override[ENV_METADATA_HOST_URL] = url
+    if token is not None:
+        config_override[ENV_METADATA_TOKEN] = token
+
+
+def get_details_from_env() -> Tuple[Optional[str], Optional[str]]:
+    host = os.environ.get(ENV_METADATA_HOST)
+    port = os.environ.get(ENV_METADATA_PORT)
+    token = os.environ.get(ENV_METADATA_TOKEN)
+    protocol = os.environ.get(ENV_METADATA_PROTOCOL, "http")
+    url = os.environ.get(ENV_METADATA_HOST_URL)
+    if port is not None:
+        url = f"{protocol}://{host}:{port}"
+        return url, token
+    # The reason for using host as URL is backward compatibility
+    # If port is not being used we assume someone is using host env var as URL
+    if url is None and host is not None:
+        logger.warning(
+            f"Do not use {ENV_METADATA_HOST} as URL. Use {ENV_METADATA_HOST_URL} instead"
+        )
+    return url or host, token
+
+
+def load_client_config() -> DatahubClientConfig:
+    try:
+        ensure_datahub_config()
+        client_config_dict = config_utils.get_client_config()
+        datahub_config = DatahubConfig.parse_obj(client_config_dict).gms
+        return datahub_config
+    except ValidationError as e:
+        click.echo(
+            f"Received error, please check your {config_utils.CONDENSED_DATAHUB_CONFIG_PATH}"
+        )
+        click.echo(e, err=True)
+        sys.exit(1)
+
+    # If config does not exist, create a default one.
+    if not datahub_config:
+        config = DatahubClientConfig()
+
+    # Override gms & token configs if specified.
+    if len(config_override.keys()) > 0:
+        config.gms_host = config_override.get(ENV_METADATA_HOST_URL)
+        config.token = config_override.get(ENV_METADATA_TOKEN)
+    elif config_utils.should_skip_config():
+        gms_host_env, gms_token_env = get_details_from_env()
+        config.gms_host = gms_host_env
+        config.token = gms_token_env
+
+    return config
+
+
+def ensure_datahub_config() -> None:
+    if not os.path.isfile(config_utils.DATAHUB_CONFIG_PATH):
+        click.secho(
+            f"No {config_utils.CONDENSED_DATAHUB_CONFIG_PATH} file found, generating one for you...",
+            bold=True,
+        )
+        write_gms_config(config_utils.DEFAULT_GMS_HOST, None)
+
+
+def write_gms_config(
+    host: str, token: Optional[str], merge_with_previous: bool = True
+) -> None:
+    config = DatahubConfig(gms=DatahubClientConfig(server=host, token=token))
+    if merge_with_previous:
+        try:
+            previous_config = config_utils.get_client_config()
+            assert isinstance(previous_config, dict)
+        except Exception as e:
+            # ok to fail on this
+            previous_config = {}
+            logger.debug(
+                f"Failed to retrieve config from file {config_utils.DATAHUB_CONFIG_PATH}: {e}. This isn't fatal."
+            )
+        config_dict = {**previous_config, **config.dict()}
```

Comment:
this merge logic is strange and probably hurts correctness

Reply:
This function already existed in `metadata-ingestion/src/datahub/cli/config_utils.py`, I simply moved it to the client.

What is your concern with correctness here? Can you provide more details?

Reply:
if it's existing behavior then we can leave it as-is. however, I do think we should be overwriting here - having the write_gms_config method not fully overwrite is highly unexpected